### PR TITLE
ORV2-1314: generate CGI files and upload to SFTP server by schedule

### PR DESCRIPTION
**Pull request by bcgov-brwang**
# Description

Please provide a summary of the change and the issue fixed. Please include relevant context. List dependency changes.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist

- [ ] I have read the [CONTRIBUTING](CONTRIBUTING.md) doc
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have already been accepted and merged

## Further comments

---

Thanks for the PR!

Deployments, as required, will be available below:
- [Frontend](https://onroutebc-1466-frontend.apps.silver.devops.gov.bc.ca)
- [Vehicles](https://onroutebc-1466-vehicles.apps.silver.devops.gov.bc.ca/api)
- [Dops](https://onroutebc-1466-dops.apps.silver.devops.gov.bc.ca/api)
- [Scheduler](https://onroutebc-1466-scheduler.apps.silver.devops.gov.bc.ca/api)

Please create PRs in draft mode.  Mark as ready to enable:
- [Analysis Workflow](https://github.com/bcgov/onroutebc/actions/workflows/analysis.yml)

After merge, new images are promoted to:
- [Merge Workflow](https://github.com/bcgov/onroutebc/actions/workflows/merge.yml)

### CODE REVIEW

File: database/mssql/scripts/versions/v_32_ddl.sql [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+IF @@ERROR <> 0 SET NOEXEC ON
+GO
+
+--------------------------------------
+-- Holiday table creation
+--------------------------------------
+CREATE TABLE [dbo].[ORBC_HOLIDAY](
+	[ID] [int] IDENTITY(1,1) NOT NULL,
+	[HOLIDAY_YEAR] [int] NOT NULL,
```
Comment (praju-aot) [inline database/mssql/scripts/versions/v_32_ddl.sql:23]: Please merge year, month, day into one single col

**Review by praju-aot — CHANGES_REQUESTED** [review]

File: scheduler/src/helper/generator.helper.ts [hunk 2]
```diff
@@ -0,0 +1,423 @@
+import { config as dotenvConfig } from "dotenv";
+import * as fs from 'fs';
+import { join, resolve } from 'path';
+import { CgiSftpService } from "src/modules/cgi-sftp/cgi-sftp.service";
+import { Transaction } from "src/modules/transactions/transaction.entity";
+import { Readable } from "typeorm/platform/PlatformTools";
+import { CgiConstants } from "src/common/constants/cgi.constant";
+
+const envFilePath = resolve(__dirname, "../../../.env");
```
Comment (krishnan-aot) [inline scheduler/src/helper/generator.helper.ts:10]: @cberg-aot This is the `generator.helper` file that was failing the other day in @praju-aot's PR build. It was deploying this image where the .env path has changed. This is a variation of the stale image.

File: scheduler/src/common/constants/cgi.constant.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+export class CgiConstants {
```
Comment (krishnan-aot) [inline scheduler/src/common/constants/cgi.constant.ts:1]: Can these variables be made into environment variables and added to vault?

Comment (krishnan-aot) [reply scheduler/src/common/constants/cgi.constant.ts:1]: Can you also add comments what each of these variables mean?

File: scheduler/src/helper/generator.helper.ts [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+    const randomChars = generateRandomChars(8); // Generate 8 random characters
+    const journalName: string = prefix + randomChars;
+    return journalName;
+  }
+  
+  function generateRandomChars(length: number): string {
+    const characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+    let result = '';
+    for (let i = 0; i < length; i++) {
+      result += characters.charAt(Math.floor(Math.random() * characters.length));
```
Comment (krishnan-aot) [inline scheduler/src/helper/generator.helper.ts:74]: use crypto.getRandom instead of Math.random.

File: scheduler/src/helper/generator.helper.ts [hunk 5]
```diff
@@ -0,0 +1,423 @@
+import { config as dotenvConfig } from "dotenv";
+import * as fs from 'fs';
+import { join, resolve } from 'path';
+import { CgiSftpService } from "src/modules/cgi-sftp/cgi-sftp.service";
+import { Transaction } from "src/modules/transactions/transaction.entity";
+import { Readable } from "typeorm/platform/PlatformTools";
+import { CgiConstants } from "src/common/constants/cgi.constant";
+
+const envFilePath = resolve(__dirname, "../../../.env");
+const result = dotenvConfig({ path: envFilePath });
```
Comment (praju-aot) [inline scheduler/src/helper/generator.helper.ts:11]: Build failure. Not required. env variables are loaded in app.module.ts.  Remove line 9 and 10

File: scheduler/src/helper/generator.helper.ts [hunk 6]
```diff
@@ -0,0 +1,423 @@
...
+    bt.transactionType = transactionType;
+    bt.fiscalYear = fiscalYear;
+    bt.batchNumber = batchNumber;
+    bt.messageVersion = messageVersion;
+    console.log(filename);
+    console.log(ackFilename);
+    return batchHeader;
+  }
+
+function populateJournalHeader(transactions: Transaction[]): string {
```
Comment (krishnan-aot) [inline scheduler/src/helper/generator.helper.ts:245]: Can you write documentation on all of these functions and provide documentation on the general flow of execution? Looking at this PR, I've no idea what any of these steps even do.

File: scheduler/src/helper/generator.helper.ts (hunk 4 above)
Comment (praju-aot) [inline scheduler/src/helper/generator.helper.ts:73]: Sonar cloud hotspot. Use cyrpto

**Review by praju-aot — CHANGES_REQUESTED** [review]

File: scheduler/src/helper/generator.helper.ts [hunk 7]
```diff
@@ -0,0 +1,423 @@
...
+
+  batchTrailer = batchTrailer + feederNumber + batchType + transactionType + delimiter + feederNumber
+    + fiscalYear + batchNumber + controlCount + controlTotal + feederNumberClientSystem + delimiter  + `\n`;
+  return batchTrailer;
+}
+
+function generateCgiFile(transactions: Transaction[]): void {
+  const now: Date = new Date();
+  const cgiCustomString: string = formatDateToCustomString(now);
+  const cgiFileName: string = `F3535.${cgiCustomString}`;
```
Comment (krishnan-aot) [inline scheduler/src/helper/generator.helper.ts:353]: What is F3535? Can you make it an env variable/constant?

File: scheduler/src/modules/transactions/holiday.entity.ts [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import { Entity, Column, PrimaryGeneratedColumn, OneToOne, JoinColumn } from 'typeorm';
+
+@Entity('ORBC_HOLIDAY', { schema: 'dbo' })
+export class BcHoliday {
+  @PrimaryGeneratedColumn()
+  ID: number;
+
+  @Column()
+  HOLIDAY_YEAR: number;
```
Comment (krishnan-aot) [inline scheduler/src/modules/transactions/holiday.entity.ts:9]: Please refer to other entity files on naming conventions.

**Review by krishnan-aot — CHANGES_REQUESTED** [review]

**Comment by praju-aot** [comment]
Kindly resolve Sonar Cloud issues reported in the below url. Code rating is C

https://sonarcloud.io/project/issues?resolved=false&sinceLeakPeriod=true&pullRequest=1466&id=onroutebc_tps_migration

**Pull request closed without merging**